README
/*
   Wampirzce - Symulacja Wampirzcow-Rozpustnikow
   Copyright (C) 2004, 2007, 2023 Krzysztof Burghardt.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/


NEWS

 Release Notes
 =============

 *******************
 1.0.1 (01 Nov 2007)
 *******************

* fixed memory allocation bug.

 *******************
 1.0.0 (27 Feb 2004)
 *******************

* first public version.

 *******************
 0.0.2 (24 Feb 2004)
 *******************

* rewritten to use SDL_opengl (not glut).
* added status bar.
* many improvments.
* small bugfix.

 *******************
 0.0.1 (04 Feb 2004)
 *******************

* rewritten using autoconf and automake.


src/main.C
/*
   Wampirzce - Symulacja Wampirzcow-Rozpustnikow
   Copyright (C) 2004 Krzysztof Burghardt.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <iostream>
#include <cstdio>
#include <SDL/SDL.h>
#include <SDL/SDL_opengl.h>
#include <GL/freeglut.h>
#include "config.h"
#include "kolonia.h"

using namespace std;

static char *id = "$Id: main.C,v 1.17 2007-11-01 17:07:20 kb Exp $";

unsigned int wielkosc_okna_x = 1024;
unsigned int wielkosc_okna_y = 768;

const unsigned int wielkosc_ekosystemu_x = 1800;
const unsigned int wielkosc_ekosystemu_y = 1400;

const unsigned int pojemnosc_srodowiska = 120;
const unsigned int wampirzcow_na_poczatku = 12;

const unsigned int jurnosc_minimalna = 2;
const unsigned int jurnosc_maksymalna = 20;

const unsigned int zycie_minimalne = 80;
const unsigned int zycie_a_jurnosc = 5;

const unsigned int odpoczynek_minimalny = 32;
const double odpoczynek_a_jurnosc = 0.05;

const unsigned int prawdopodobienstwo_mutacji = 20;	// w %
const double prawdopodobienstwo_mutacji_a_jurnosc = 0.5;

const double szybkosc_a_jurnosc = 0.7;

unsigned int
  Wampirzec::ilosc_wampirzcow = 0;

Kolonia *
  KoloniaWampirzcow;

int
main (int argc, char **argv)
{
  cout << PACKAGE_STRING << endl << id << endl;

  SDL_Init (SDL_INIT_VIDEO | SDL_INIT_TIMER);
  glutInit (&argc, argv);
  SDL_GL_SetAttribute (SDL_GL_DOUBLEBUFFER, true);
  SDL_GL_SetAttribute (SDL_GL_RED_SIZE, 5);
  SDL_GL_SetAttribute (SDL_GL_GREEN_SIZE, 5);
  SDL_GL_SetAttribute (SDL_GL_BLUE_SIZE, 5);
  SDL_Surface *
    surface = SDL_SetVideoMode (wielkosc_okna_x,
				wielkosc_okna_y,
				16,
				SDL_OPENGL | SDL_ANYFORMAT | SDL_DOUBLEBUF
				| SDL_FULLSCREEN);
  if(!surface)
    {
	clog << "BUG: " << __FILE__ << ":" << __LINE__ <<                                                                                    
            " main() - SDL_SetVideoMode() nie utworzylo powierzchni!" << endl;
	return 0;
    }
  
  SDL_WM_SetCaption (PACKAGE_STRING, PACKAGE_STRING);

  SDL_Event
    event;
  bool
    go_quit = false;
  unsigned int
    startTime = 0;
  unsigned int
    endTime = 0;

  KoloniaWampirzcow = new Kolonia;

  while (!go_quit)
    {
      startTime = SDL_GetTicks ();
      while (SDL_PollEvent (&event))
	{
	  switch (event.type)
	    {
	    case SDL_KEYDOWN:
	      break;
	    case SDL_KEYUP:
	      switch (event.key.keysym.sym)
		{
		case SDLK_r:	// restart symulacji
		  delete KoloniaWampirzcow;
		  KoloniaWampirzcow = new Kolonia;
		  break;
		case SDLK_ESCAPE:	// koniec symulacji
		case SDLK_q:
		case SDLK_w:
		  go_quit = true;
		  break;
		}
	      break;
	    case SDL_QUIT:
	      go_quit = true;
	      break;
	    }
	}
      glMatrixMode (GL_PROJECTION);
      glLoadIdentity ();
      glOrtho (0.0, wielkosc_ekosystemu_x, 0,
	       wielkosc_ekosystemu_y - ((wielkosc_ekosystemu_y / 50) + 5),
	       -1.0, 1.0);
      glClearColor (0.0, 0.0, 0.0, 0.0);
      glClear (GL_COLOR_BUFFER_BIT);
      KoloniaWampirzcow->ChwilaZycia ();
      glFlush ();
      SDL_GL_SwapBuffers ();
      endTime = SDL_GetTicks ();
      if (endTime - startTime < 5)
	SDL_Delay (5 - (startTime - endTime));
    }

  delete
    KoloniaWampirzcow;
  SDL_Quit ();
  return 0;
}


src/kolonia.C
/*
   kolonia.C - implementacja klasy kolonia

   This file is part of Wampirzce.
   Copyright (C) 2004 Krzysztof Burghardt.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

// $Id: kolonia.C,v 1.8 2007-11-01 17:07:20 kb Exp $

#include <iostream>
#include <SDL/SDL.h>
#include <SDL/SDL_opengl.h>
#include <GL/glut.h>
#include "kolonia.h"

using namespace std;

static void
glDrawText (const char *text)
{
  for (unsigned int i = 0; i < strlen (text); i++)
    {
      glutStrokeCharacter (GLUT_STROKE_MONO_ROMAN, text[i]);
    }
}

Kolonia::Kolonia (void)
{
  iteracji = 0;
  for (unsigned int i = 0; i < wampirzcow_na_poczatku; ++i)
    {
      stado.push_back (new Wampirzec ());
    }
  (*stado.begin())->UstawIloscWampirzcow (wampirzcow_na_poczatku);

  return;
}

Kolonia::~Kolonia (void)
{
  list < Wampirzec * >::iterator poczatek = stado.begin ();
  list < Wampirzec * >::iterator koniec = stado.end ();

  while (poczatek != koniec)
    {
      delete (*poczatek);
      ++poczatek;
    }

  stado.empty ();

  return;
}

void
Kolonia::ChwilaZycia (void)
{
  ++iteracji;
  list < Wampirzec * >::iterator poczatek = stado.begin ();
  list < Wampirzec * >::iterator koniec = stado.end ();
  Wampirzec *nowy = NULL;

  if (poczatek == koniec)
    {
      cout << "Wszystkie Wampirzce wyginely..." << endl;
      exit (0);
    }

  while (poczatek != koniec)
    {
      if ((*poczatek)->CzyMartwy ())
	{
	  stado.erase (poczatek++);
	  continue;
	}
      if ((nowy = (*poczatek++)->ChwilaZycia (stado)))
	{
	  stado.push_back (nowy);
	}
    }

  glMatrixMode (GL_PROJECTION);
  glLoadIdentity ();
  glOrtho (0.0, wielkosc_okna_x, 0,
	   wielkosc_okna_y - ((wielkosc_okna_y / 50.0) + 5), -1.0, 1.0);
  glColor4f (0.7f, 0.6f, 0.5f, 0.4f);
  glEnable (GL_BLEND);
  glBlendFunc (GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
  glBegin (GL_QUADS);
  glVertex2f (0.f, 0.f);
  glVertex2f (wielkosc_okna_x, 0.f);
  glVertex2f (wielkosc_okna_x, 20.f);
  glVertex2f (0.f, 20.f);
  glEnd ();

  glColor3f (1.f, 1.f, 1.f);
  glEnable (GL_LINE_SMOOTH);
  glHint (GL_LINE_SMOOTH_HINT, GL_NICEST);

  char buffer[32];
  glScalef (0.1f, 0.1f, 1.0f);

  glTranslatef (100.f, 50.f, 0.f);
  glDrawText ("Wampirzcow: ");
  sprintf (buffer, "%d", stado.size ());
  glDrawText (buffer);

  glTranslatef (wielkosc_okna_x / 2 + 10, 0.f, 0.f);
  glDrawText ("Iteracji: ");
  sprintf (buffer, "%d", iteracji);
  glDrawText (buffer);

  glTranslatef (wielkosc_okna_x * 3.0 / 4.0 + 10, 0.f, 0.f);
  glDrawText ("Klawiszologia: [R]estart, [W]yjscie");

  glDisable (GL_LINE_SMOOTH);

  return;
}


src/kolonia.h
/*
   kolonia.h - interfejs klasy kolonia

   This file is part of Wampirzce.
   Copyright (C) 2004 Krzysztof Burghardt.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

// $Id: kolonia.h,v 1.6 2007-11-01 17:07:20 kb Exp $

#ifndef __KOLONIA_H__
#define __KOLONIA_H__

#include <list>
#include "wampirzec.h"

using namespace std;

extern unsigned int wielkosc_okna_x;
extern unsigned int wielkosc_okna_y;
extern const unsigned int wielkosc_ekosystemu_x;
extern const unsigned int wielkosc_ekosystemu_y;
extern const unsigned int wampirzcow_na_poczatku;

class Kolonia
{
public:
  Kolonia (void);
   ~Kolonia (void);

  void ChwilaZycia (void);
private:
    list < Wampirzec * >stado;
  unsigned long long int iteracji;
};

#endif


src/polozenie.h
/*
   polozenie.h - interfejs abstrakcyjnej klasy polozenie

   This file is part of Wampirzce.
   Copyright (C) 2004 Krzysztof Burghardt.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

// $Id: polozenie.h,v 1.3 2007-11-01 17:07:20 kb Exp $

#ifndef __POLOZENIE_H__
#define __POLOZENIE_H__

using namespace std;

extern const unsigned int wielkosc_ekosystemu_x;
extern const unsigned int wielkosc_ekosystemu_y;

class Polozenie
{
protected:
  Polozenie (const double, const double);
  double PobierzPolozenieX (void);
  double PobierzPolozenieY (void);
  void UstawPolozenieXY (const double, const double);
private:
  double x;
  double y;
};

#endif


src/wampirzec.h
/*
   wampirzec.h - interfejs klasy wampirzec

   This file is part of Wampirzce.
   Copyright (C) 2004 Krzysztof Burghardt.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

// $Id: wampirzec.h,v 1.5 2007-11-01 17:07:20 kb Exp $

#ifndef __WAMPIRZEC_H__
#define __WAMPIRZEC_H__

#include <list>
#include <sys/time.h>
#include "polozenie.h"

using namespace std;

static inline int
rnd (void)
{
  static struct timeval tv;
  gettimeofday (&tv, 0);
  srandom (tv.tv_sec * tv.tv_usec);
  return random () % (RAND_MAX + 1);
}

#define getrandom(min, max) ((rnd() % (int)(((max)+1) - (min))) + (min))

extern const unsigned int wielkosc_ekosystemu_x;
extern const unsigned int wielkosc_ekosystemu_y;
extern const unsigned int pojemnosc_srodowiska;
extern const unsigned int jurnosc_minimalna;
extern const unsigned int jurnosc_maksymalna;
extern const unsigned int zycie_minimalne;
extern const unsigned int zycie_a_jurnosc;
extern const unsigned int odpoczynek_minimalny;
extern const double odpoczynek_a_jurnosc;
extern const unsigned int prawdopodobienstwo_mutacji;
extern const double prawdopodobienstwo_mutacji_a_jurnosc;
extern const double szybkosc_a_jurnosc;

class Wampirzec:private Polozenie
{
public:
  Wampirzec (const double _x = getrandom (0, wielkosc_ekosystemu_x),
	     const double _y = getrandom (0, wielkosc_ekosystemu_y),
	     const double _jurnosc =
	     getrandom (jurnosc_minimalna, jurnosc_maksymalna),
	     const unsigned int _zycie =
	     zycie_minimalne + getrandom (0, zycie_minimalne),
	     const unsigned int _odpoczynek = odpoczynek_minimalny);
  Wampirzec *ChwilaZycia (const list < Wampirzec * >);
  bool CzyMartwy (void);
  void UstawIloscWampirzcow (const unsigned int);
protected:
  void Przesun (const unsigned char, const double);
  double Odleglosc (Wampirzec *);
  void Odpocznij (void);
  Wampirzec *SzukajPartnera (list < Wampirzec * >);
  Wampirzec *RozmnazajSie (Wampirzec *);
private:
  double jurnosc;
  unsigned int zycie;
  unsigned int odpoczynek;
  static unsigned int ilosc_wampirzcow;
};

#endif


src/wampirzec.C
/*
   wampirzec.C - implementacja klasy wampirzec

   This file is part of Wampirzce.
   Copyright (C) 2004 Krzysztof Burghardt.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

// $Id: wampirzec.C,v 1.9 2007-11-01 17:07:20 kb Exp $

#include <iostream>
#include <cmath>
#include <SDL/SDL.h>
#include <SDL/SDL_opengl.h>
#include "wampirzec.h"

using namespace std;

inline void
glPutPixel (double x, double y, double z, double r, double g, double b)
{
  glColor3f (r, g, b);
  glRectf (x - z, y + z, x + z, y - z);
}

Wampirzec::Wampirzec (const double _x, const double _y, const double _jurnosc,
		      const unsigned int _zycie,
		      const unsigned int _odpoczynek):
Polozenie (_x, _y)
{
  jurnosc = _jurnosc;
  zycie = _zycie + (int) (zycie_a_jurnosc * jurnosc);
  odpoczynek = _odpoczynek - (int) (odpoczynek_a_jurnosc * jurnosc);

  return;
}

Wampirzec *
Wampirzec::ChwilaZycia (const list < Wampirzec * >wszystkie_wampirzce)
{
  Wampirzec *nowy = NULL;
  if (zycie)
    {
      glPutPixel (PobierzPolozenieX (), PobierzPolozenieY ()
		  - ((wielkosc_ekosystemu_y / 50) + 5), 5,
		  1, jurnosc / 20, jurnosc / 10);
      if (odpoczynek)
	{
	  Odpocznij ();
	}
      else
	{
	  Wampirzec *ZnalezionyPartner = NULL;
	  if ((ZnalezionyPartner = SzukajPartnera (wszystkie_wampirzce)) != 0)
	    {
	      nowy = RozmnazajSie (ZnalezionyPartner);
	    }
	}
      --zycie;
    }

  return nowy;
}

bool
Wampirzec::CzyMartwy (void)
{
  if (zycie <= 0)
    {
      --ilosc_wampirzcow;
      return 1;
    }
  else
    {
      return 0;
    }
}

void
Wampirzec::UstawIloscWampirzcow (const unsigned int nowa_ilosc_wampirzcow)
{
  ilosc_wampirzcow = nowa_ilosc_wampirzcow;
}

void
Wampirzec::Przesun (unsigned char kierunek = 5, const double _odleglosc = 1)
{
  double odleglosc = _odleglosc + getrandom (0, szybkosc_a_jurnosc * jurnosc);
  switch (kierunek)
    {
    case 1:
      UstawPolozenieXY (PobierzPolozenieX () - odleglosc,
			PobierzPolozenieY () - odleglosc);
      break;
    case 2:
      UstawPolozenieXY (PobierzPolozenieX (),
			PobierzPolozenieY () - odleglosc);
      break;
    case 3:
      UstawPolozenieXY (PobierzPolozenieX () + odleglosc,
			PobierzPolozenieY () - odleglosc);
      break;
    case 4:
      UstawPolozenieXY (PobierzPolozenieX () - odleglosc,
			PobierzPolozenieY ());
      break;
    case 5:
      UstawPolozenieXY (PobierzPolozenieX (), PobierzPolozenieY ());
      break;
    case 6:
      UstawPolozenieXY (PobierzPolozenieX () + odleglosc,
			PobierzPolozenieY ());
      break;
    case 7:
      UstawPolozenieXY (PobierzPolozenieX () - odleglosc,
			PobierzPolozenieY () + odleglosc);
      break;
    case 8:
      UstawPolozenieXY (PobierzPolozenieX (),
			PobierzPolozenieY () + odleglosc);
      break;
    case 9:
      UstawPolozenieXY (PobierzPolozenieX () + odleglosc,
			PobierzPolozenieY () + odleglosc);
      break;
    default:
      clog << "BUG: " << __FILE__ << ":" << __LINE__ <<
	" Wampirzec::Przesun() - nieznany kierunek!" << endl;
    }

  return;
}

double
Wampirzec::Odleglosc (Wampirzec * partner)
{
  return sqrt (pow (PobierzPolozenieX () - partner->PobierzPolozenieX (), 2)
	       + pow (PobierzPolozenieY () - partner->PobierzPolozenieY (),
		      2));
}

void
Wampirzec::Odpocznij (void)
{
  Przesun (getrandom (1, 9), getrandom (1, 3));
  --odpoczynek;
  return;
}

Wampirzec *
Wampirzec::SzukajPartnera (list < Wampirzec * >cale_stado)
{
  Wampirzec *najblizszy = NULL;
  list < Wampirzec * >::iterator poczatek = cale_stado.begin ();
  list < Wampirzec * >::iterator koniec = cale_stado.end ();
  najblizszy = (*poczatek++);

  while (poczatek != koniec)
    {
      if (*poczatek != this &&	// sam z soba sie nie rozmnozy
	  Odleglosc (*poczatek) < Odleglosc (najblizszy))
	{
	  najblizszy = (*poczatek);
	}
      ++poczatek;
    }

  if (Odleglosc (najblizszy) <= 3)	// sex na odleglosc? ;-P
    {
      return najblizszy;
    }
  else
    {
      unsigned char kierunek = getrandom (1, 9);
      if (PobierzPolozenieX () < najblizszy->PobierzPolozenieX ())
	{
	  if (PobierzPolozenieY () < najblizszy->PobierzPolozenieY ())
	    {
	      kierunek = 9;
	    }
	  else if (PobierzPolozenieY () == najblizszy->PobierzPolozenieY ())
	    {
	      kierunek = 6;
	    }
	  else
	    {
	      kierunek = 3;
	    }
	}
      else if (PobierzPolozenieX () == najblizszy->PobierzPolozenieX ())
	{
	  if (PobierzPolozenieY () < najblizszy->PobierzPolozenieY ())
	    {
	      kierunek = 8;
	    }
	  else
	    {
	      kierunek = 2;
	    }
	}
      else
	{
	  if (PobierzPolozenieY () < najblizszy->PobierzPolozenieY ())
	    {
	      kierunek = 7;
	    }
	  else if (PobierzPolozenieY () == najblizszy->PobierzPolozenieY ())
	    {
	      kierunek = 4;
	    }
	  else
	    {
	      kierunek = 1;
	    }
	}
      Przesun (kierunek);
    }

  return NULL;
}

Wampirzec *
Wampirzec::RozmnazajSie (Wampirzec * partner)
{
  Wampirzec *nowy = NULL;
  if (partner != this && ilosc_wampirzcow <= pojemnosc_srodowiska)
    {
      // dojdzie do mutacji kodu genetycznego?
      if (getrandom (0, 100) <
	  prawdopodobienstwo_mutacji -
	  (prawdopodobienstwo_mutacji_a_jurnosc * jurnosc))
	{
	  nowy = new Wampirzec (PobierzPolozenieX (), PobierzPolozenieY ());
	}
      else
	{			// nie, dziedzicza po jednym z rodzicow
	  double j = getrandom (0, 1) ? jurnosc : partner->jurnosc;
	  unsigned int z = getrandom (0, 1) ? zycie : partner->zycie;
	  unsigned int o =
	    getrandom (0, 1) ? odpoczynek : partner->odpoczynek;
	  nowy =
	    new Wampirzec (PobierzPolozenieX (), PobierzPolozenieY (), j, z,
			   o);
	}
      ++ilosc_wampirzcow;
    }

  odpoczynek =
    odpoczynek_minimalny +
    (int) (odpoczynek_minimalny * odpoczynek_a_jurnosc * jurnosc);
  partner->odpoczynek =
    odpoczynek_minimalny +
    (int) (odpoczynek_minimalny * odpoczynek_a_jurnosc * jurnosc);

  return nowy;
}
